[zdn][new] 506352 - Guess the Mass

Kalendar2 opened the pull request.

Kalendar2 committed:
Файл 506352.js добавлен в папку

SugarHedgehog commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,74 @@
+let ind = [0, 1, 2, 3];
+let ind2 = [0, 1, 2, 3];
+
+function shuffle(array) {
```
Такая функция уже присутствует во внутренней библиотеке.

SugarHedgehog commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,74 @@
+let rand = sluchch(0, 10);
```
Не используемая переменная

SugarHedgehog commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,74 @@
+let rand = sluchch(0, 10);
+let m = 'масса';
```
Переменные должны отображать то, для чего они предназначенны. 
Если вам нужна строка 'масса', то положите её в перемнную `weight` или на край транслитом `massa`
Всех следующих переменных это тоже касается

SugarHedgehog commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,74 @@
+array3[3] = big[bi];
+
+let i = 0;
+while (i != 4) {
```
Для этого стоит использовать for. В идеале map

SugarHedgehog commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,74 @@
+}
+
+i = 4;
+while (i != 0) {
```
Вот это точно map

SugarHedgehog reviewed (commented):
Не подключено к main.js

Kalendar2 committed:
Файл 506352.js изменён

Kalendar2 committed:
Файл main.js добавлен в соответсвующую папку

SugarHedgehog commented on zdn/matege2024b/2/506352.js:
```diff
@@ -9,22 +8,11 @@ let small = ['монеты', 'малины', 'клубники', 'вилки', '
+index2.shuffle();
 
 let array3 = [];
 let v = sluchch(0, verysmall.length - 1);
```
Категорически не понимаю, зачем вы создаёте 4 следующие переменные и тем более не понимаю зачем вы их связываете с будущим весом объектов (тут бы sl() хватило). Определите, как уже опытный программист, `array3` в одну строку и не создавайте лапши из использования переменных.

SugarHedgehog commented on zdn/matege2024b/2/506352.js:
```diff
@@ -9,22 +8,11 @@ let small = ['монеты', 'малины', 'клубники', 'вилки', '
-
-shuffle(ind);
-shuffle(ind2);
+index1.shuffle();
```
Зачем перемешивать потом, когда можно сразу?

SugarHedgehog commented on zdn/matege2024b/2/506352.js:
```diff
@@ -42,33 +30,29 @@ array3[1] = small[s];
-while (i != 0) {
-    i--;
-    ans[i] = ans[i] + 1;
+for (let i = 3; i >= 0; i--) {
```
map вам в помощь

SugarHedgehog reviewed (commented):
Оптимизируйте и причешите код, пожалуйста

Kalendar2 committed:
Файл 506352.js изменён

SugarHedgehog commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,53 @@
+let me = sluchch(0, medium.length - 1);
+let bi = sluchch(0, big.length - 1);
+
+a = (v + 1) * sluchch(10, 20);
```
Хочу получить обоснование такой инициализации веса
Как, индекс выбранного вами объекта, связан с его весом? Изменится ли что-то идейно, если определять 
`a = sl(1,7) * sluchch(10, 20);` или проще `a = sl(10,140);`

nickkolok commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,53 @@
+let c;
+let d;
+let verysmall = ['таблетки лекарства', 'комара', 'мухи', 'божьей коровки', 'капли воды', 'иголки', 'пуговицы'];
+let small = ['монеты', 'малины', 'клубники', 'вилки', 'яйца', 'ножниц', 'птицы', 'компьютерной мыши', 'картофеля'];
```
Хотел бы я, чтобы выращенная у меня на даче малина весила как картофелина... Тут главное - не перепутать, что к чему приравнивать.

Kalendar2 committed:
В файле 506292.js изменены названия переменных

Kalendar2 committed:
В файле 506292.js добавлены новые объекты для повышения разнообразия генерируемых задач

Kalendar2 committed:
В файле 506352.js изменён способ инициализации веса, а также добавлены несколько новых объектов в соответсвующии массивы

nickkolok commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,54 @@
+let bigObject = [['машины', 'фонарного столба'].iz(), 'морского ската', ['бегемота', 'трактора'].iz(), 'носорога', 'индийского слона', 'касатки', 
+['грузовика', 'бульдозера'].iz(), 'африканского слона', 'автобуса', 'фуры', 'гусенечного трактора', 'танка'];
+
+let sluchIndexVerySmallObject = sluchch(0, verySmallObject.length - 1);
```
Ага! То есть объекты в массиве расположены не в случайном порядке, а в порядке пропорционального возрастания диапазона масс? Интересное решение! Однако у него есть некоторые проблемы. Например, этот же код предполагается использовать для задач на линейные величины - например, в таких задачах может попасться глкбина Байкала, которая 1637 метров безо всякой случайной компоненты. А может попасться рост человека, который смело может быть от 1.5 м до 2 м.

Поэтому я предлагаю реорганизовать всю эту конструкцию примерно так.

```js
let mediumObject = [
    ['холодильника', sl(20,200)],
    ['человека', sl(47,120)],
    ['книжного шкафа', sl(15,100)],
];
```

Если есть возражения или идеи, как сделать лучше - не стесняйтесь высказать!

Кстати, @SugarHedgehog , вот и ответ на вопрос, зачем так странно генерить случайные числа: через произведение.

SugarHedgehog replied:
>Кстати, @SugarHedgehog , вот и ответ на вопрос, зачем так странно генерить случайные числа: через произведение.

Мне кажется, на этот вопрос должен был ответить @Kalendar2... Но ладно

Kalendar2 committed:
В файле 506352.js исправлен баг, при котором число тонн иногда имело более одного знака после запятой

nickkolok commented:
Хорошо, но замечания исправлены не все. Продолжайте в том же духе!

Kalendar2 committed:
mass заменён на weight, а также оптимизирован и сделан более читабельным код

nickkolok commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,121 @@
+	objects.map(([name, weightFunction]) => [name, weightFunction()]);
+
+// Определяем объекты с их весами
+let verySmallObjects = createObjectsWithWeights([
```
Выглядит гораздо лучше!
Однако всё ещё непонятно, зачем в массиве хранить аж целую стрелочную функцию вместо того, чтобы просто хранить три аргумента `sl()` числами?

nickkolok commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,121 @@
+
+let answer = `A) ${arrayAnswer[0]} B) ${arrayAnswer[1]} C) ${arrayAnswer[2]} D) ${arrayAnswer[3]}`;
+
+NAtask.setTask({
```
Неплохо бы тут увидеть формулировку задания!

nickkolok commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,121 @@
+
+NAtask.setTask({
+	text: ` A) ${weightWord} ${arraySluchObjects[0]} B) ${weightWord} ${arraySluchObjects[1]} ` +
+		`C) ${weightWord} ${arraySluchObjects[2]} D) ${weightWord} ${arraySluchObjects[3]} $$ $$ ` +
```
А двойной доллар зачем?

nickkolok commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,121 @@
+	parseFloat((weightT).toFixed(1)) + ' т'
+];
+
+let answer = `A) ${arrayAnswer[0]} B) ${arrayAnswer[1]} C) ${arrayAnswer[2]} D) ${arrayAnswer[3]}`;
```
Ответ, если мне не изменяет память, записывается вовсе не так, а последовательностью цифр без пробелов.

nickkolok commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,121 @@
+	weightMg + ' мг',
+	weightG + ' г',
+	weightKg + ' кг',
+	parseFloat((weightT).toFixed(1)) + ' т'
```
А чем тонна такая особенная?

nickkolok replied:
Ожидается ответ!

Kalendar2 committed:
Стрелочная функция убрана из массива, добавлена формулировка задания, изменён формат ответа, изменено отображение задания

Kalendar2 committed:
для тонны добавлена функция ts(), для сокращения и большей выразительности кода

nickkolok commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,116 @@
+	weightMg + ' мг',
+	weightG + ' г',
+	weightKg + ' кг',
+	weightT.ts() + ' т'
```
Чем тонна такая особенная, что для неё нужен `.ts()` ? Что сломается, если его убрать?

nickkolok replied:
Ответ на вопрос всё ещё ожидается!

nickkolok commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,116 @@
+let answer = arrayAnswer[0] + '' + arrayAnswer[1] + '' + arrayAnswer[2] + '' + arrayAnswer[3];
+
+NAtask.setTask({
+	text: ` Установите соответствие между величинами и их возможными значениями: ` +
```
Зачем начинать текст задания с пробела?

nickkolok commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,116 @@
+NAtask.setTask({
+	text: ` Установите соответствие между величинами и их возможными значениями: ` +
+		`к каждому элементу первого столбца подберите соответствующий элемент из второго столбца.` +
+		`$$ A) \\ ${weightWord} \\ ${arraySluchObjects[0]} \\\\ B) \\ ${weightWord} \\ ${arraySluchObjects[1]} \\\\ ` +
```
Каков глубокий смысл помещать ответы в $\LaTeX$ ? На первый взгляд - одни сплошные неудобства! Пробелы, например, приходится экранировать.

nickkolok commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,116 @@
+	weightT.ts() + ' т'
+];
+
+let answer = arrayAnswer[0] + '' + arrayAnswer[1] + '' + arrayAnswer[2] + '' + arrayAnswer[3];
```
Зачем "прокладывать" между строками пустые строки? Почему вообще не использовать `.join()` ?

nickkolok commented on zdn/matege2024b/2/main.js:
```diff
@@ -0,0 +1,3 @@
+window.nomer=[
+	506352,
+].iz();
```
А `comment` ?

Kalendar2 committed:
Убран пробел перед текстом, добавлена join, ответы теперь не помещаются в Латех и записаны в два столбца

Kalendar2 committed:
Добавлен комментарий в main.js

nickkolok commented:
Если Вы внимательно посмотрите вот сюда:
https://github.com/nickkolok/chas-ege/pull/2743/files#diff-b5384aa24fbe847a2a08efc5e0f109c44b258bd26dabc04fc845dff1c09ed25b

то Вы увидите, чт у нас появилась новая функция-обёртка: `setCorrespondenceTask`. (Да, Вы настолько долго и упорно трудитесь над этим шаблоном!). А если Вы пересоберётесь с `origin/devel`, то там появится эта функция... наряду с подрежкой смешанных чисел, которую Вы добавили.

nickkolok commented:
Ещё одна проблема - что иногда бывает так, что с одной величиной измерения две позиции:

<img width="917" height="621" alt="изображение" src="https://github.com/user-attachments/assets/16aa1296-c949-479b-9740-3471d0528352" />


При этом они должны достаточно сильно отличаться (во сколько раз как минимум?).

Kalendar2 committed:
Добавлена новая обёртка и возможность генерации двух одинаковых позициий

nickkolok commented:
Во-первых, переход на `setCorrespondenceTask` должны был быть отдельным коммитом, а всё остальное - другим коммитом.

Во-вторых, а как контролируется то, чтобы не выпало две очень похожих (или даже одинаковых) массы одновременно?

В-третьих, прошу применить стратегическое мышление. Ведь на базе этого кода Вам предстоит писать и задачки про длину. объём и т.д. Он должен быть расширяемым!

Kalendar2 commented:
<img width="588" height="286" alt="image" src="https://github.com/user-attachments/assets/811cc925-7303-4f3a-9e7b-0942d1245784" />
2) Если один из объектов берётся из  около начала массива, то с шансом 50/50 выбирается значение приблизительно из конца массива. Значения достаточно сильно отличаются по массе, чтобы не спутать их
3) Не совсем понимаю, что можно изменить в коде. Ведь для задач про длину и объём, достаточно будет изменить только значения и названия объектов

Kalendar2 committed:
Добавлены ещё 4 меры, добавлена функция getSelectedPreferenceFromList, четвёртая категория может случайным образом переводиться в значения третьей, сохраняя свои размеры (например 2 дня в 48 часов, 3 километра в 3000 метров и т.д.). К сожалению не успел к концу недели закончить сами массивы с объектами и их диапозонами значений, но в скором времени закончу

Kalendar2 committed:
Закончены массивввы объектов, добавлена проверка значений на их близость друг к другу, добавлены минуты для явлений измеряющихся в часах, куб.м и кв.м заменены на м в степени

Kalendar2 committed:
Merge branch 'devel' of github.com:nickkolok/chas-ege into newtask-506352

Kalendar2 committed:
Merge branch 'newtask-506352' of github.com:kalendar2/chas-ege into newtask-506352

Kalendar2 committed:
Добавлена проверка на дубликаты

nickkolok committed:
[zdn][fix] 506352: toBaseCoeff — площадь 1e-12→1e-6, объём 1e-9→1e-6

1 мм² = (1e-3 м)² = 1e-6 м² (было 1e-12 — занижено в 10⁶ раз)
1 мл  = 1 см³ = (1e-2 м)³ = 1e-6 м³ (было 1e-9 — занижено в 10³ раз)
Из-за этого сортировка valuesInBase и проверка «отличаются в 5 раз»
для case 3 и case 4 работали некорректно.

nickkolok committed:
[zdn][fix] 506352: minutes → timeSuffix, явная конкатенация

Было: глобальная minutes с лидирующими пробелами (" 15 мин"),
      молчаливо приклеивалась к measurements[2].
Стало: timeSuffix без пробелов, пробел добавляется в тернарнике
       только если суффикс непустой.

nickkolok committed:
[zdn][fix] 506352: убран «вес», оставлена только «масса»

Масса (кг, г, мг, т) и вес (Н) — разные физические величины.
Случайная подстановка «вес» при значениях в граммах — методическая ошибка.

nickkolok committed:
[zdn][codestyle] 506352 - remove trailing spaces

nickkolok commented:
https://chat.qwen.ai/s/t_a2f281e6-6ca9-40cf-9d27-fba78a7ce5d3?fev=0.2.80

Qwen-3.8-Max-Preview

Пока так, дальше, честно говоря, пока не знаю, что делать.

nickkolok committed:
[zdn][fix] 506352 - deduplicate buses

SugarHedgehog commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,493 @@
+    valueS * toBaseCoeff[unitIndices[1]],
+    valueM * toBaseCoeff[unitIndices[2]],
+    valueB * toBaseCoeff[unitIndices[3]]
+].sort((a, b) => a - b);
```
Это sortNumeric

SugarHedgehog commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,493 @@
+		solution: arrayValue[indexValue[i]]
+	}));
+
+	let right = indexObject.map(i => arrayValue[i]);
```
Это копирование?

SugarHedgehog commented on zdn/matege2024b/2/506352.js:
```diff
@@ -0,0 +1,493 @@
+	let mediumObjects = createObjectsWithValues(objectsArray3);
+	let bigObjects = createObjectsWithValues(objectsArray4);
+
+	let indexObject = [0, 1, 2, 3];
```
Для какой цели этот блок?

SugarHedgehog replied:
setCorrespondenceTask вроде сам мешает отлично

SugarHedgehog reviewed (commented):
Разделить на блоки комментариями с объяснениями

nickkolok committed:
[zdn][ref] 506352: sortNumeric(), убрать ручное перемешивание, комментарии

- .sort((a,b)=>a-b) заменён на .sortNumeric() (из lib/array_general.js)
- Убраны indexObject/indexValue и их .shuffle():
  setCorrespondenceTask сам перемешивает left и right
- left строится напрямую из arrayObjects
- right = arrayValue.slice() вместо indexObject.map(...)
- Добавлены комментарии к основным блокам кода

Замечания SugarHedgehog от 26.07.2026

nickkolok commented:
https://chat.qwen.ai/s/t_376ca1e9-800e-487f-bc0e-352e62038da9?fev=0.2.80
